### PR TITLE
content_encoding: support concatenated gzip members

### LINKED ISSUE #22156: --compressed fails on multi-member (concatenated) gzip responses
### I did this

I was fetching large compressed log archives over HTTP with `curl --compressed`.
Some of these are made by concatenating several gzip files, which is a common shape for rotated logs and what `pigz` or a plain `cat a.gz b.gz` produce. 
Per RFC 1952 section 2.2 that is still one valid gzip stream (gunzip and zcat decode it fine), but curl stops partway through.

curl decodes only the first gzip member and then fails the transfer with `CURLE_WRITE_ERROR (23)`, dropping the rest of the data.

Minimal reproduction, no real server needed:

```python
import zlib
data = b"HTTP/1.1 200 OK\r\nContent-Encoding: gzip\r\nConnection: close\r\n\r\n"
def member(s):
    c = zlib.compressobj(9, zlib.DEFLATED, zlib.MAX_WBITS | 16)
    return c.compress(s) + c.flush()
data += member(b"foo") + member(b"bar")   # two concatenated gzip members
open("foobar.bin", "wb").write(data)
```

```
# terminal 1
while true; do nc -l 127.0.0.1 8000 < foobar.bin; done
# terminal 2
curl -sv --compressed -o out http://127.0.0.1:8000/
```

`out` ends up with `foo` and curl exits with code 23. `gunzip < foobar.bin`
returns `foobar`.

### I expected the following

`out` to contain `foobar` and the transfer to succeed, the same way gunzip, zcat
and `gzip.decompress()` in Python handle it.

### Notes

This came up before in #3796, which went stale without a conclusion. 
That thread weighed two options: silently ignore everything after the first member (what some browsers do), or decode every member. The argument against silently ignoring was data integrity, since curl is usually an automated tool and quietly losing data
is easy to miss.

Decoding all members keeps that integrity and matches the reference tools. 
The spot is `process_trailer()` in `lib/content_encoding.c`: once zlib returns `Z_STREAM_END` for the first member, the remaining bytes (the next member, starting with the gzip magic `1f 8b`) are taken as unexpected trailing data and rejected. The usual zlib idiom is to call `inflateReset()` when `Z_STREAM_END` is hit with input still pending and continue.

I have a patch and a test and can send a PR if this direction looks right.

### curl/libcurl version

Reproduced with curl 8.7.1 (zlib/1.2.12). The relevant code in `lib/content_encoding.c` is unchanged on current master.

### operating system

macOS, but it is not platform specific.

### PR DESCRIPTION
**Pull request by vigneshwarrvenkat**
Fixes #22156

## Summary
- Fix `--compressed` / `Content-Encoding: gzip` handling for concatenated gzip members (RFC 1952 section 2.2)
- On `Z_STREAM_END`, reset inflate for the next member instead of calling `exit_zlib()` early
- Added regression test `tests/data/test2217` with two gzip members (`foo` + `bar` → `foobar`)
Fixes #22156
## Problem
When a response contains multiple gzip members (common for rotated logs / `pigz` output), curl decoded only the first member and failed with `CURLE_WRITE_ERROR (23)`.
## Solution
Add gzip-specific member-end handling in `lib/content_encoding.c`:
- If the next bytes are gzip magic (`0x1f 0x8b`), `inflateReset2()` and continue
- If no input remains, reset state and wait for more body data (cleanup in `gzip_do_close()`)
- Leave deflate/raw-deflate paths unchanged
## Test plan
- [x] `./tests/runtests.pl -v 220 2217`
- [x] `make checksrc`

### CODE REVIEW

File: lib/content_encoding.c [hunk 1]
```diff
@@ -144,6 +144,29 @@ static CURLcode process_trailer(struct Curl_easy *data, struct zlib_writer *zp)
...
+    zp->zlib_init = ZLIB_INIT_GZIP;
+    *keep_going = TRUE;
+    return CURLE_OK;
+  }
+  if(z->avail_in)
+    return exit_zlib(data, z, &zp->zlib_init, CURLE_WRITE_ERROR);
+  if(inflateReset2(z, MAX_WBITS + 32) != Z_OK)
+    return exit_zlib(data, z, &zp->zlib_init, process_zlib_error(data, z));
+  zp->zlib_init = ZLIB_INIT_GZIP;
+  return CURLE_OK;
```
[the remark this answers is not in the document]
Comment (bagder) [reply lib/content_encoding.c:167]: This seems like an accurate observation about an edge case that probably is hard to make a test case for...

**Comment by HendrikLambert** [comment]
Hi @bagder @vigneshwarrvenkat I saw commit df350dd added an explicit error for multi-member gzip streams. We encounter concatenated gzip data at DuckDB, so we're still very interested in seeing this supported natively.

If no one is actively working on it, I'd be happy to pick up where this left off and address what Copilot flagged. Let me know if I should pick it up and do so via a new PR or contribute to this one.

**Comment by vigneshwarrvenkat** [comment]
@HendrikLambert I'm working on it. Will raise a PR soon.

**Comment by vigneshwarrvenkat** [comment]
@HendrikLambert Did this PR address your problem https://github.com/curl/curl/issues/22156

**Comment by HendrikLambert** [comment]
@vigneshwarrvenkat Yes, our problem is exactly the same, concatenated gzip streams from log ingestion tools that set the Content-Encoding: gzip header. This PR solves these cases.

**Comment by jay** [comment]
> * On `Z_STREAM_END`, reset inflate for the next member instead of calling `exit_zlib()` early

Isn't the server at fault here, where does it say we are supposed to do this? When I [researched](https://github.com/curl/curl/issues/3796#issuecomment-485338006) it I found Firefox and Chrome ignored the additional members.

/cc @monnerat

**Comment by HendrikLambert** [comment]
According to [RFC 1952, Section 2.2 (File format)](https://datatracker.ietf.org/doc/html/rfc1952#section-2.2):

> A gzip file consists of a series of "members" (compressed data sets). The format of each member is specified in the following section. The members simply appear one after another in the file, with no additional information before, between, or after them.

This part of the gzip specification is, for instance, being heavily used by log ingestion tools that upload to S3. This allows them to compress and upload each part separately and concatenate them afterwards, still supplying valid gzip. 

GNU Gzip also supports this, as do most language standard libraries (Go, Python)

**Comment by jay** [comment]
I know Stefan gave an explanation of this and I don't want to butcher it but basically my understanding is the server appears to be wrong because it didn't encode the file. The server sees the gzip extension on the archive and is probably like ok I'll serve this with content encoding gzip and that's where I think it's wrong. What would be technically ok but not make a lot of sense is if it encoded the gzip file with gzip then it could add the content encoding header because it encoded it.

**Comment by HendrikLambert** [comment]
I completely understand where Stefan and you are coming from, but the server is actually following standard HTTP practices here.

Using `Content-Encoding: gzip` does not require the server to compress the data dynamically on the fly (although commonly done). Serving a pre-compressed `.gz` file directly from disk with a `Content-Encoding: gzip` header is used extensively by CDNs, S3 and can also be enabled in Nginx with `gzip_static on;`. The header simply tells the client, "The payload over the wire is gzip-compressed, please decompress it before handing it to the application."

i.e. client request `foo.txt` with `Accept-Encoding: gzip` -> server looks for pre-compressed `foo.txt.gz` and responds with `Content-Encoding: gzip` (and `Content-Type: text/plain`). The server could also pre-compute `foo.txt.zstd`, etc.

Since the payload being served is a valid multi-member gzip stream (per RFC 1952) and the server correctly flagged it as gzip with `Content-Encoding`, the client should fully decompress all members of the stream.

The same happens with `zstd` under the hood. It also supports multiple frames, but `libzstd` deals with this automatically, so curl inherits that.

**Comment by icing** [comment]
I understand why people want to update already gzipped, large log files by adding new members. What I a missing here is the documentation under which conditions these members are decompressed into a single file?

This is not mentioned in the specification. It mentions a *filename*  for each member, supposedly extracting the content into separate files. But that is not possible when removing the `Content-Encoding`  from a HTTP response. So, exactly when should this be done while preventing misuse by attackers that attach their payload to such archives?

To repeat here, so it does not get lost: curl will transfer those gzip files just fine. Only when asked to strip the content encoding, it does not what you want it to do, right?

**Comment by HendrikLambert** [comment]
When using gzip as a stream compressor, the `FLG.FNAME` is not set. As the compression is done over the file content, it does not hold any file-specific metadata. Furthermore, I don't even think GNU Gzip supports the creation of multiple files in a single gzip, as all directory information must be stripped per the Gzip spec. This is why tar is used to bundle the files, where the resulting stream is then compressed with gzip.

Since `Content-Encoding: gzip` strictly focuses on gzip as a stream encoding, not tar + gzip as an archive format, gzip is compressing *one* file. Adding more members to a `.gz` file just makes said file longer, it does not add more files. Log ingestion tools use exactly this semantic. 

To expand upon my earlier example:
A log ingestion tool creates a single file `log.txt.gz` file, where it periodically appends a new gzip member (no `FNAME`)
Client requests `log.txt` with `Accept-Encoding: gzip` -> web server responds with the the content of `log.txt.gz` with `Content-Encoding: gzip` and `Content-Type: text/plain`, which is seen as one file (`log.txt`) by the client.

To confirm your summary: curl CLI will transfer the raw, compressed payload when you *dont* use `--compressed` and add the `Content-Encoding: gzip` header yourself. When using `--compressed`, curl stops after the first member since df350dd with an error message.

**Comment by dfandrich** [comment]
> but the server is actually following standard HTTP practices here.

I guess you mean to say "common" HTTP practices as opposed to "standard", which
has a different meaning.

**Comment by monnerat** [comment]
> Isn't the server at fault here, where does it say we are supposed to do this? When I https://github.com/curl/curl/issues/3796#issuecomment-485338006 it I found Firefox and Chrome ignored the additional members.

First, I'm neither a guru of this, nor the author of this decoding, I just adapted it to support multiple decoding stages.
But IMO, this is the big difference between Transfer-Encoding and Content-Encoding: The former describes the output stream while the latter is an info about the data before the server processes it.

This is a non-sense to have server-generated multiple members (Transfer-Encoding) while effective data can be a real archive in this case, use options to not decode, as this is not the role of curl to behave as an archive handler.

**Comment by HendrikLambert** [comment]
To ensure we're all aligned on the underlying semantics and I'm not missing anything:

> Unlike Content-Encoding ([Section 8.4.1](https://www.rfc-editor.org/info/rfc9110/#section-8.4.1) of [[HTTP](https://www.rfc-editor.org/info/rfc9112/#HTTP)]), Transfer-Encoding is a property of the message, not of the representation. Any recipient along the request/response chain MAY decode the received transfer coding(s) or apply additional transfer coding(s) to the message body
— *[RFC 9112: HTTP/1.1 - 6.1. Transfer-Encoding](https://www.rfc-editor.org/info/rfc9112/#section-6.1:~:text=Unlike,body%2C)*

My interpretation: `Transfer-Encoding` is on a message (hop-by-hop) basis, while `Content-Encoding` says something about the representation (end-to-end). `Transfer-Encoding` is also not part of the HTTP/2 spec, so in HTTP/2 the only way to do compression without the application itself being involved is `Content-Encoding`. `Content-Type` describes the media type associated with the *uncompressed/original* representation.

@monnerat  
> This is a non-sense to have server-generated multiple members (Transfer-Encoding) while effective data can be a real archive in this case, use options to not decode, as this is not the role of curl to behave as an archive handler.

If the server is sending an actual archive, it will serve `Content-Type: application/x-tar` alongside `Content-Encoding: gzip`. If it serves `Content-Type: text/plain` with `Content-Encoding: gzip`, the application expects a decoded text file, not an archive. Curl's job is simply to remove the `Content-Encoding` (by fully inflating the stream) before handing the text/plain payload to the application, which is currently not being done. Only the first member is being inflated.

The fact that GZip (without an archiving format like tar) has options in the spec to include a name per member does make this somewhat ambiguous, but the `FNAME` field is optional exactly for the reason to use gzip as a stream compressor, without any knowledge of the underlying representation, such as file names or metadata.

Precedence for this can also be found in `zlib` itself. The `gzread()` function automatically looks for new members after `Z_STREAM_END` https://github.com/madler/zlib/blob/e3dc0a85b7032e98380dec011bc8f2c2ee0d8fca/gzread.c#L231 and completely ignores the `FNAME` field if present. This way of dealing with a gzip stream is also the case in libarchive, Go (compress/gzip), Python (gzip) and most likely more, but this is what I verified.

**Comment by icing** [comment]
> HTTP/2 spec, so in HTTP/2 the only way to do compression without the application itself being involved is `Content-Encoding`. `Content-Type` describes the media type associated with the _uncompressed/original_ representation.

A server can serve the `gzip` resource simply without a `Content-Encoding`. You can transfer any data via HTTP. The setup abuses `Content-Encoding` for what it is not intended for. Many server offer to send "pre-compressed" resources from static files and slap a `Content-Encoding` on it, which saves CPU. That is a good optimisation. However, the feature intention is to serve the same bits as if the encoding were applied "on-the-fly", as a singe-entry gzip.

Now, the multi-entry gzips for log files make sense for the easy of updates. It's just treating this as a content encoding result is not what HTTP clients expect. Just transfer the compressed files without that header and everything is fine.
